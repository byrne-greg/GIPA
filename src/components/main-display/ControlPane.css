.ControlPane {
  display: flex;
  justify-content: space-between;
  align-items: center;
  flex-direction: column;
}

.ControlPane__inputPane {
  width: 100%;
  display: flex;
  justify-content: space-between;
  align-items: center;
  flex-direction: row;
}

.InputField {
  width: 100vw;
  margin: 10px 0 10px 10px;
  padding: 12px;
  font-size: calc(10px + 2vmin);
  box-sizing: border-box;
  border-radius: 4px;
  border: 2px solid grey;
}

.InputField:focus {
  border: 2px solid black;
  outline: none;
}

.SubmitButton {
  border:none;
  margin:0 5px;
  color: #73348E;
  background-color: white;
  font-size: calc(30px + 2vmin);
  cursor: pointer;
}

.InputButton {
  border: solid;
  margin: 10px;
  padding: 10px;
  color: #73348E;
  background-color: white;
  font-size: calc(30px + 2vmin);
  cursor: pointer;
 
}

.SuggestedResponsePane {
  display: flex;
  justify-content: space-around;
  align-items: left;
  flex-flow: column wrap;
}

.SuggestedResponsePane__title {
  margin: 5px 10px;
  text-align: left;
  color: grey;
  text-decoration: underline;
}

.SuggestedResponsePane__title p {
  font-size: calc(8px + 2vmin);
  font-style: italic;  
}

.SuggestedResponsePane__responses {
  display: flex;
  justify-content: space-around;
  align-items: center;
  flex-flow: row wrap;
}
