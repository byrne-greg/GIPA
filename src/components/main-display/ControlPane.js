import "./ControlPane.css";
import React from "react";
import PropTypes from "prop-types";
import InfoPane from "./InfoPane";
import shortid from "shortid";

export default class ControlPane extends React.Component {
  
	constructor(props) {
		super(props);

		this.state = {inputValue: ""};
		this.handleChange = this.handleChange.bind(this);
		this.handleSubmit = this.handleSubmit.bind(this);
		this.handleInputButton = this.handleInputButton.bind(this);
		this.handleKeyPress = this.handleKeyPress.bind(this);
	}
  
	enterMessage(messageInput) {
		this.props.handleEnteredMessage(messageInput);
		document.getElementById("inputField").value = "";
		this.setState({inputValue: ""});
	}
  
	handleKeyPress(event) {
		if (event.key === "Enter") {
			this.enterMessage(event.target.value);
		}
	}
 
	handleSubmit() {
		this.enterMessage(this.state.inputValue);    
	}

	handleInputButton(messageInput) {
		this.props.handleEnteredMessage(messageInput);

	}

	handleChange(event) {
		document.getElementById("inputField").scrollIntoView({ behavior: "smooth"});
		this.setState({inputValue: event.target.value});
	}

	render() {
		return (      
			<div className="ControlPane">
				<SuggestedResponsePane suggestedResponses={this.props.watsonResponseSuggestions} handleSubmit={this.handleInputButton}/>
				<div className="ControlPane__inputPane">
					<InputField handleChange={this.handleChange} handleKeyPress={this.handleKeyPress}/>
					<SubmitButton handleSubmit={this.handleSubmit} hasInputText={this.state.inputValue !== ""}/>
					<InfoPane>
						<SuggestedResponsePane suggestedResponses={["Tell me a joke", "Does Greg have friends?", "Do you have a partner", "What are Greg's kids?", "Where does he work?", "What is Greg"]} handleSubmit={this.handleInputButton}/>
					</InfoPane>
				</div>
			</div>
		);
	}
}
ControlPane.propTypes = {
	handleEnteredMessage: PropTypes.func.isRequired,
	watsonButtonSuggestions: PropTypes.arrayOf(PropTypes.string.isRequired)
};
ControlPane.defaultProps = {
	suggestedResponses: []
};

export const InputField = (props) => {
	return(
		<input 
			className="InputField"
			id="inputField" 
			type="text" placeholder="Ask me something..."
			onChange={props.handleChange}
			onKeyPress={props.handleKeyPress} 
		/>
	);
};
InputField.propTypes = {
	handleChange: PropTypes.func.isRequired,
	handleKeyPress: PropTypes.func.isRequired
};

export const SubmitButton = (props) => {
	return(
		<button className="SubmitButton" onClick={props.handleSubmit} style={props.hasInputText ? {opacity:1} : {opacity:0.5}}><i className="fas fa-comment"></i></button>
	);
};
SubmitButton.propTypes = {
	handleSubmit: PropTypes.func.isRequired,
	hasInputText: PropTypes.bool
};
SubmitButton.defaultProps = {
	hasInputText: false
};

export const InputButton = (props) => {
	return (      
		<button className="InputButton" onClick={() => props.handleSubmit(props.buttonText)}><p>{props.buttonText}</p></button>
	);
};
InputButton.propTypes = {
	handleSubmit: PropTypes.func.isRequired,
	buttonText: PropTypes.string
};
InputButton.defaultProps = {
	buttonText: "Hi"
};

export const SuggestedResponsePane = (props) => {
	return (
		<div className="SuggestedResponsePane">
			<div className="SuggestedResponsePane__title">
				<p>GIPA suggests asking:</p>
			</div>
			<div className="SuggestedResponsePane__responses">
				{props.suggestedResponses.map(response => (<InputButton key={shortid.generate()} handleSubmit={props.handleSubmit} buttonText={response}/>))}
			</div>
		</div>
	);
};
SuggestedResponsePane.propTypes = {
	handleSubmit: PropTypes.func.isRequired,
	suggestedResponses: PropTypes.arrayOf(PropTypes.string.isRequired),
};
SuggestedResponsePane.defaultProps = {
	suggestedResponses: []
};

