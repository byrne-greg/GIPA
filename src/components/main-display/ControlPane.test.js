import React from "react";
import ControlPane, {InputField, SubmitButton, InputButton, SuggestedResponsePane} from "./ControlPane.js";
import renderer from "react-test-renderer";
import {shallow} from "enzyme";


describe("ControlPane component", () => {

	it("shallow renders without crashing", () => {
		shallow(<ControlPane handleEnteredMessage={() => {}}/>);
	});

	it("full render matches expected snapshot", () => {
		const controlPane = renderer.create(<ControlPane handleEnteredMessage={() => {}}/>).toJSON();
		expect(controlPane).toMatchSnapshot();
	});

});

describe("SubmitButton component", () => {

	it("shallow renders without crashing", () => {
		var wrapper = shallow(<SubmitButton handleSubmit={() => {}}/>);
		expect(wrapper.find("button").length).toBe(1);
	});

	it("shallow renders with default opacity", () => {
		var wrapper = shallow(<SubmitButton handleSubmit={() => {}}/>);    
		expect(wrapper.find("button").prop("style")).toEqual({opacity:0.5});
	});

	it("shallow renders with half opacity when no input text exists", () => {
		var wrapper = shallow(<SubmitButton handleSubmit={() => {}} hasInputText={false} />);    
		expect(wrapper.find("button").prop("style")).toEqual({opacity:0.5});
	});

	it("shallow renders with full opacity when input text exists", () => {
		var wrapper = shallow(<SubmitButton handleSubmit={() => {}} hasInputText={true} />);    
		expect(wrapper.find("button").prop("style")).toEqual({opacity:1});
	});

});

describe("InputButton component", () => {

	it("shallow renders without crashing", () => {
		let buttonText="test";
		var wrapper = shallow(<InputButton handleSubmit={() => {}} buttonText={buttonText}/>);
		expect(wrapper.find("button").length).toBe(1);
		expect(wrapper.find("button").text()).toEqual(buttonText);
	});

});

describe("SuggestedResponsePane component", () => {

	it("shallow renders without crashing", () => {
		var wrapper = shallow(<SuggestedResponsePane handleSubmit={() => {}}/>);
		expect(wrapper.find(".SuggestedResponsePane").length).toBe(1);
	});

});

describe("InputField component", () => {

	it("shallow renders without crashing", () => {
		var wrapper = shallow(<InputField handleChange={() => {}} handleKeyPress={() => {}}/>);
		expect(wrapper.find("input").length).toBe(1);
	});

});