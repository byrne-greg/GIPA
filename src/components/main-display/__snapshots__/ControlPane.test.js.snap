// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlPane component full render matches expected snapshot 1`] = `
<div
  className="ControlPane"
>
  <div
    className="SuggestedResponsePane"
  >
    <div
      className="SuggestedResponsePane__title"
    >
      <p>
        GIPA suggests asking:
      </p>
    </div>
    <div
      className="SuggestedResponsePane__responses"
    />
  </div>
  <div
    className="ControlPane__inputPane"
  >
    <input
      className="InputField"
      id="inputField"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Ask me something..."
      type="text"
    />
    <button
      className="SubmitButton"
      onClick={[Function]}
      style={
        Object {
          "opacity": 0.5,
        }
      }
    >
      <i
        className="fas fa-comment"
      />
    </button>
    <div
      className="InfoPane"
    >
      <button
        className="InfoPane__overlayOpenButton"
        onClick={[Function]}
      >
        <i
          className="fas fa-info-circle"
        />
      </button>
      <div
        className="InfoPane__overlay"
        style={
          Object {
            "display": "none",
          }
        }
      >
        <div
          className="Card__content"
        >
          <h2
            className="Card__title"
          >
            Greg's Interactive Personal Advocate
          </h2>
          <div
            className="Card__body"
          >
            <p>
              GIPA was created as a personal project. The idea was to devise a way in which inquiring users could ask any questions they want to find out more about me. It's real mission was a self-learning practical exercise in using a wide variety of unfamiliar technologies and experimenting with my favourite topic at present time of writing - Chatbots!. Read about it's 
              <a
                href="https://medium.com/@byrne.greg/a-development-story-in-creating-a-digital-advocate-46237c7be8bb"
                rel="noopener noreferrer"
                target="_blank"
              >
                development story
              </a>
               or visit the 
              <a
                href="https://github.com/byrne-greg/GIPA"
                rel="noopener noreferrer"
                target="_blank"
              >
                code repository
              </a>
              .
            </p>
            <p>
              GIPA uses a bunch of technologies. The front-end was developed with 
              <a
                href="https://reactjs.org/"
                rel="noopener noreferrer"
                target="_blank"
              >
                ReactJS
              </a>
               (and typical CSS/HTML). The conversational agent was created using the 
              <a
                href="https://www.ibm.com/watson/ai-assistant/"
                rel="noopener noreferrer"
                target="_blank"
              >
                IBM's Watson Assistant
              </a>
               platform. The connectivity between the front-end and Watson Assistant uses the 
              <a
                href="https://samnewman.io/patterns/architectural/bff/"
                rel="noopener noreferrer"
                target="_blank"
              >
                Backend for Frontend
              </a>
               pattern. It is deployed using a 
              <a
                href="https://nodejs.org/en/"
                rel="noopener noreferrer"
                target="_blank"
              >
                NodeJS
              </a>
               FaaS with 
              <a
                href="https://openwhisk.apache.org/"
                rel="noopener noreferrer"
                target="_blank"
              >
                Apache OpenWhisk
              </a>
              . The Backend for Frontend FaaS and the static front-end content is all hosted on 
              <a
                href="https://www.ibm.com/cloud/"
                rel="noopener noreferrer"
                target="_blank"
              >
                IBM Cloud
              </a>
               (formerly Bluemix). The VCS uses 
              <a
                href="https://github.com/"
                rel="noopener noreferrer"
                target="_blank"
              >
                GitHub
              </a>
               and the build toolchain uses 
              <a
                href="https://travis-ci.org/"
                rel="noopener noreferrer"
                target="_blank"
              >
                Travis CI
              </a>
               and 
              <a
                href="https://coveralls.io/"
                rel="noopener noreferrer"
                target="_blank"
              >
                Coveralls
              </a>
              .
            </p>
            <p>
              For functionality, the responses are hard-coded within the Assistant workspace (due to the mostly static information about me). There is some dynamic content such as asking for a joke, which is a call-out to 
              <a
                href="https://icanhazdadjoke.com/"
                rel="noopener noreferrer"
                target="_blank"
              >
                https://icanhazdadjoke.com/
              </a>
               to retrieve a random joke.
            </p>
            <p>
              Try ask GIPA questions in natural language about me.
            </p>
            <div />
          </div>
        </div>
        <div
          className="InfoPane__suggestedResponse"
          onClick={[Function]}
        >
          <div
            className="SuggestedResponsePane"
          >
            <div
              className="SuggestedResponsePane__title"
            >
              <p>
                GIPA suggests asking:
              </p>
            </div>
            <div
              className="SuggestedResponsePane__responses"
            >
              <button
                className="InputButton"
                onClick={[Function]}
              >
                <p>
                  Tell me a joke
                </p>
              </button>
              <button
                className="InputButton"
                onClick={[Function]}
              >
                <p>
                  Does Greg have friends?
                </p>
              </button>
              <button
                className="InputButton"
                onClick={[Function]}
              >
                <p>
                  Do you have a partner
                </p>
              </button>
              <button
                className="InputButton"
                onClick={[Function]}
              >
                <p>
                  What are Greg's kids?
                </p>
              </button>
              <button
                className="InputButton"
                onClick={[Function]}
              >
                <p>
                  Where does he work?
                </p>
              </button>
              <button
                className="InputButton"
                onClick={[Function]}
              >
                <p>
                  What is Greg
                </p>
              </button>
            </div>
          </div>
        </div>
        <div
          className="CloseButton"
          onClick={[Function]}
        >
          <i
            class="far fa-times-circle"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
